hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,325 @@
+		// separate loop to ensure all STAKED_TO_ME fields have their final values
+		updateNodeStakes(pendingChanges);
+		updateFundingRewardBalances(pendingChanges);
+
+		super.preview(pendingChanges);
+
+		if (!rewardsActivated) {
+			checkStakingRewardsActivation(pendingChanges);
+		}
+	}
+
+	private void updateAccountStakes(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		for (int i = 0, n = pendingChanges.size(); i < n; i++) {
+			hasBeenRewarded[i] = false;
+
+			final var account = pendingChanges.entity(i);
+			final var changes = pendingChanges.changes(i);
+
+			// Update BALANCE and STAKE_PERIOD_START in the pending changes for this account, if reward-eligible
+			if (isRewardable(account, changes)) {
+				payReward(i, account, changes);
+			}
+			// Update any STAKED_TO_ME side effects of this change
+			n = updateStakedToMeSideEffects(account, changes, pendingChanges);
+		}
+	}
+
+	private void updateFundingRewardBalances(
+			final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		final var rewardsPaid = rewardCalculator.rewardsPaidInThisTxn();
+		if (rewardsPaid > 0) {
+			final var rewardAccountI = stakeChangeManager.findOrAdd(800L, pendingChanges);
+			updateBalance(-rewardsPaid, rewardAccountI, pendingChanges);
+		}
+	}
+
+	private void updateNodeStakes(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		for (int i = 0, n = pendingChanges.size(); i < n; i++) {
+			final var account = pendingChanges.entity(i);
+			final var changes = pendingChanges.changes(i);
+
+			final var curNodeId = (account != null) ? account.getStakedId() : 0L;
+			final var newNodeId = getNodeStakeeNum(changes);
+			if (curNodeId < 0 && curNodeId != newNodeId) {
+				// Node stakee has been replaced, withdraw stakeRewarded or stakeNotRewarded from ex-stakee based on
+				// isDeclineReward option
+				stakeChangeManager.withdrawStake(
+						account.getStakedNodeAddressBookId(), // since nodeId is saved as  -nodeId -1
+						account.getBalance() + account.getStakedToMe(),
+						account != null && account.isDeclinedReward());
+			}
+			if (newNodeId < 0) {
+				// Award updated stake to new node stakee to the fields stakeRewarded or stakeNotRewarded from
+				// ex-stakee based on isDeclineReward option
+				stakeChangeManager.awardStake(
+						Math.abs(newNodeId + 1), // since nodeId is saved as  -nodeId -1
+						finalBalanceGiven(account, changes) + finalStakedToMeGiven(account, changes),
+						finalDeclineRewardGiven(account, changes));
+			}
+		}
+	}
+
+	int updateStakedToMeSideEffects(
+			final MerkleAccount account,
+			final Map<AccountProperty, Object> changes,
+			final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		int changesSize = pendingChanges.size();
+
+		final var curStakeeNum = (account != null) ? account.getStakedId() : 0L;
+		final var newStakeeNum = getAccountStakeeNum(changes);
+
```

Comment:
:warning: I think here we should insert,
```suggestion
if (curStakeeNum == newStakeeNum) {
    return changesSize;
}
```
...since _neither_ `if` below should happen if the stakee hasn't really changed.